code/ION4.1.1/ici/doc/man/man3/lyst.3
.\" Automatically generated by Pod::Man v1.37, Pod::Parser v1.32
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "pod3::lyst 3"
.TH pod3::lyst 3 "2022-05-20" "perl v5.8.8" "ICI library functions"
.SH "NAME"
lyst \- library for manipulating generalized doubly linked lists
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 1
\&    #include "lyst.h"
.Ve
.PP
.Vb 2
\&    typedef int  (*LystCompareFn)(void *s1, void *s2);
\&    typedef void (*LystCallback)(LystElt elt, void *userdata);
.Ve
.PP
.Vb 1
\&    [see description for available functions]
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
The \*(L"lyst\*(R" library uses two types of objects, \fILyst\fR objects
and \fILystElt\fR objects.  A Lyst knows how many elements it contains, 
its first and last elements, the memory manager used
to create/destroy the Lyst and its elements, and how the elements are
sorted.  A LystElt knows its content (normally a pointer to an item
in memory), what Lyst it belongs to, and the LystElts before and after
it in that Lyst.
.IP "Lyst lyst_create(void)" 4
.IX Item "Lyst lyst_create(void)"
Create and return a new Lyst object without any elements in it.
All operations performed on this Lyst will use the
allocation/deallocation functions of the default memory
manager \*(L"std\*(R" (see \fImemmgr\fR\|(3)).  Returns \s-1NULL\s0 on any failure.
.IP "Lyst lyst_create_using(unsigned memmgrId)" 4
.IX Item "Lyst lyst_create_using(unsigned memmgrId)"
Create and return a new Lyst object without any elements in it.
All operations performed on this Lyst will use the
allocation/deallocation functions of the specified
memory manager (see \fImemmgr\fR\|(3)).  Returns \s-1NULL\s0 on any failure.
.IP "void lyst_clear(Lyst list)" 4
.IX Item "void lyst_clear(Lyst list)"
Clear a Lyst, i.e. free all elements of \fIlist\fR, calling the Lyst's
deletion function if defined, but without destroying the Lyst itself.
.IP "void lyst_destroy(Lyst list)" 4
.IX Item "void lyst_destroy(Lyst list)"
Destroy a Lyst.  Will free all elements of \fIlist\fR, calling the Lyst's
deletion function if defined.
.IP "void lyst_compare_set(Lyst list, LystCompareFn compareFn)" 4
.IX Item "void lyst_compare_set(Lyst list, LystCompareFn compareFn)"
.PD 0
.IP "LystCompareFn lyst_compare_get(Lyst list)" 4
.IX Item "LystCompareFn lyst_compare_get(Lyst list)"
.PD
Set/get comparison function for specified Lyst.  Comparison 
functions are called with two Lyst element data
pointers, and must return a negative integer if first
is less than second, 0 if both are equal, and a positive integer
if first is greater than second (i.e., same return values as \fIstrcmp\fR\|(3)).
The comparison function is used by the
\&\fIlyst_insert()\fR, \fIlyst_search()\fR, \fIlyst_sort()\fR, and \fIlyst_sorted()\fR
functions.
.IP "void lyst_direction_set(Lyst list, LystSortDirection direction)" 4
.IX Item "void lyst_direction_set(Lyst list, LystSortDirection direction)"
Set sort direction (either \s-1LIST_SORT_ASCENDING\s0 or
\&\s-1LIST_SORT_DESCENDING\s0) for specified Lyst.  If no comparison
function is set, then this controls whether
new elements are added to the end or beginning (respectively) 
of the Lyst when \fIlyst_insert()\fR is called.
.IP "void lyst_delete_set(Lyst list, LystCallback deleteFn, void *userdata)" 4
.IX Item "void lyst_delete_set(Lyst list, LystCallback deleteFn, void *userdata)"
Set user deletion function for specified Lyst.  This
function is automatically called whenever an element of the Lyst is deleted,
to perform any user-required processing.  When automatically called,
the deletion function is passed two arguments: the element being deleted
and the \fIuserdata\fR pointer specified in the \fIlyst_delete_set()\fR call.
.IP "void lyst_insert_set(Lyst list, LystCallback insertFn, void *userdata)" 4
.IX Item "void lyst_insert_set(Lyst list, LystCallback insertFn, void *userdata)"
Set user insertion function for specified Lyst.  This
function is automatically called whenever a Lyst element is
inserted into the Lyst, to perform any user-required processing.
When automatically called, the insertion function is passed two arguments:
the element being inserted and the \fIuserdata\fR pointer specified in
the \fIlyst_insert_set()\fR call.
.IP "unsigned long lyst_length(Lyst list)" 4
.IX Item "unsigned long lyst_length(Lyst list)"
Return the number of elements in the Lyst.
.IP "LystElt lyst_insert(Lyst list, void *data)" 4
.IX Item "LystElt lyst_insert(Lyst list, void *data)"
Create a new element whose content is the pointer value \fIdata\fR
and insert it into the Lyst.  Uses the Lyst's comparison
function to select insertion point, if defined; otherwise
adds the new element at the beginning or end of the Lyst,
depending on the Lyst sort direction setting.  Returns a
pointer to the newly created element, or \s-1NULL\s0 on any failure.
.IP "LystElt lyst_insert_first(Lyst list, void *data)" 4
.IX Item "LystElt lyst_insert_first(Lyst list, void *data)"
.PD 0
.IP "LystElt lyst_insert_last(Lyst list, void *data)" 4
.IX Item "LystElt lyst_insert_last(Lyst list, void *data)"
.PD
Create a new element and insert it at the beginning/end
of the Lyst.  If these functions are used when inserting elements
into a Lyst with a defined comparison function, then the Lyst may
get out of order and future calls to \fIlyst_insert()\fR can put new elements 
in unpredictable locations.  Returns a pointer to
the newly created element, or \s-1NULL\s0 on any failure.
.IP "LystElt lyst_insert_before(LystElt element, void *data)" 4
.IX Item "LystElt lyst_insert_before(LystElt element, void *data)"
.PD 0
.IP "LystElt lyst_insert_after(LystElt element, void *data)" 4
.IX Item "LystElt lyst_insert_after(LystElt element, void *data)"
.PD
Create a new element and insert it before/after the
specified element.  If these functions are used when inserting
elements into a Lyst with a defined comparison function,
then the Lyst may get out
of order and future calls to \fIlyst_insert()\fR can put new
elements in unpredictable locations.  Returns a pointer
to the newly created element, or \s-1NULL\s0 on any failure.
.IP "void lyst_delete(LystElt element)" 4
.IX Item "void lyst_delete(LystElt element)"
Delete the specified element from its Lyst and deallocate its memory.  
Calls the user delete function if defined.
.IP "LystElt lyst_first(Lyst list)" 4
.IX Item "LystElt lyst_first(Lyst list)"
.PD 0
.IP "LystElt lyst_last(Lyst list)" 4
.IX Item "LystElt lyst_last(Lyst list)"
.PD
Return a pointer to the first/last element of a Lyst.
.IP "LystElt lyst_next(LystElt element)" 4
.IX Item "LystElt lyst_next(LystElt element)"
.PD 0
.IP "LystElt lyst_prev(LystElt element)" 4
.IX Item "LystElt lyst_prev(LystElt element)"
.PD
Return a pointer to the element following/preceding the specified element.
.IP "LystElt lyst_search(LystElt element, void *searchValue)" 4
.IX Item "LystElt lyst_search(LystElt element, void *searchValue)"
Find the first matching element in a Lyst starting with
the specified element.  Returns \s-1NULL\s0 if no matches are
found.  Uses the Lyst's comparison function if defined,
otherwise searches from the given element to the end of the Lyst.
.IP "Lyst lyst_lyst(LystElt element)" 4
.IX Item "Lyst lyst_lyst(LystElt element)"
Return the Lyst to which the specified element belongs.
.IP "void* lyst_data(LystElt element)" 4
.IX Item "void* lyst_data(LystElt element)"
.PD 0
.IP "void* lyst_data_set(LystElt element, void *data)" 4
.IX Item "void* lyst_data_set(LystElt element, void *data)"
.PD
Get/set the pointer value content of the specified Lyst element.  The
set routine returns the element's previous content, and the
delete function is \fInot\fR called.  If the \fIlyst_data_set()\fR
function is used on an element of a Lyst with a defined comparison
function, then the Lyst may get out of order and future calls to
\&\fIlyst_insert()\fR can put new elements in unpredictable locations.
.IP "void lyst_sort(Lyst list)" 4
.IX Item "void lyst_sort(Lyst list)"
Sort the Lyst based on the current comparison function
and sort direction.  A stable insertion sort is used
that is very fast when the elements are already in order.
.IP "int lyst_sorted(Lyst list)" 4
.IX Item "int lyst_sorted(Lyst list)"
Determine whether or not the Lyst is sorted based on
the current comparison function and sort direction.
.IP "void lyst_apply(Lyst list, LystCallback applyFn, void *userdata)" 4
.IX Item "void lyst_apply(Lyst list, LystCallback applyFn, void *userdata)"
Apply the function \fIapplyFn\fR automatically to each element
in the Lyst.  When automatically called, \fIapplyFn\fR is passed
two arguments: a pointer to an element, and the \fIuserdata\fR
argument specified in the call to \fIlyst_apply()\fR.  \fIapplyFn\fR
should not delete or reorder the elements in the Lyst.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fImemmgr\fR\|(3), \fIpsm\fR\|(3)


code/ION4.1.1/ici/library/lyst.c
/*
 * lyst.c -- subroutines for managing a doubly linked list ("lyst").
 *
 *	Copyright (c) 1997, California Institute of Technology.
 *	ALL RIGHTS RESERVED.  U.S. Government Sponsorship
 *	acknowledged.
 *	Author: Jeff Biesiadecki, Jet Propulsion Laboratory
 *	Adapted by Scott Burleigh, Jet Propulsion Laboratory
 *
 * $Log: lyst.c,v $
 * Revision 1.3  2007/07/07 19:38:39  scott
 * Revamp ICI.
 *
 * Revision 1.2  2006/05/21 16:20:50  scott
 * Add llcv to ici, clean up header inclusion.
 *
 * Revision 1.1.1.1  2004/12/28 04:21:39  scott
 * Original source
 *
 * Revision 1.1.1.1  2004/12/06 16:37:31  scott
 * Original code base
 *
 * Revision 1.2  2002/12/17 16:14:25  scott
 * Added interface to sptrace, to enable tracing of DRAM and SDR space usage.
 *
 * Revision 1.1.1.1  2002/04/26 17:29:14  scott
 * CFDP as a DTN application
 *
 * Revision 1.1.1.1  2002/03/13 01:08:25  scott
 * CFDP and supporting stuff
 *
 * Revision 1.5  2001/2/19  20:30:00  scott
 * Extracted memory management abstraction into 'memmgr' library.
 *
 * Revision 1.4  1997/10/11  10:00:00  scott
 * Added symbolic naming of memory allocation/deallocation function pairs.
 *
 * Revision 1.3  1995/05/16  23:50:00  scott
 * Added ability to specify memory deallocation function as well as
 * allocation function.
 *
 * Revision 1.2  1995/03/28  22:14:11  scott
 * Add ability to select memory allocation function when creating a list.
 *
 * Revision 1.1  1994/05/12  20:56:32  chuck
 * Initial revision
 *
 * Revision 1.8  1994/04/13  20:17:17  jeffb
 * split lyst_compare_set/get into two functions
 * lyst_insert will insert either last or first when no compare function,
 * depending on how list->dir is set
 *
 * Revision 1.7  1994/04/13  20:08:16  jeffb
 * add lyst_compare_get function
 *
 * Revision 1.6  1994/03/03  00:33:09  jeffb
 * change insert/delete callbacks to refer to element that was inserted
 * or was about to be deleted
 *
 * Revision 1.5  1994/03/02  01:11:37  jeffb
 * added insert function
 *
 * Revision 1.4  1994/02/16  04:13:03  jeffb
 * changed mind - take out not too useful sorted flag
 *
 * Revision 1.3  1994/02/16  03:56:50  jeffb
 * always sort when lyst_sort called even if list->sorted is true
 *
 * Revision 1.2  1994/02/16  02:28:18  jeffb
 * replace LystPointer with void *
 * remove assert() statements from checking user provided parameters
 * add "sorted" flag and function
 *
 * Revision 1.1  1994/02/15  06:24:29  jeffb
 * Initial revision
 *
 */

#ifdef __ghs__
#define NDEBUG
#endif

#include "platform.h"
#include "memmgr.h"
#include "lystP.h"

/*
 * prototypes for private functions
 */

static void lyst__clear(Lyst);
static int lyst__inorder(Lyst,void *,void *);
static LystElt lyst__elt_create(const char *, int, Lyst, void *);
static void lyst__elt_clear(LystElt);
static char *lyst__alloc(const char *, int, int, unsigned int);
static void lyst__free(const char *, int, int, char *);

/*
 * public functions -- create and destroy list objects
 */

Lyst
Lyst_create(const char *file, int line)
{
  return Lyst_create_using(file, line, 0);
}

Lyst
Lyst_create_using(const char *file, int line, int idx)
{
  Lyst list;

  if ((list = (Lyst) lyst__alloc(file, line, idx, sizeof(*list))) == NULL)
  {
	putErrmsg("Can't create list.", NULL);
	return NULL;
  }

  lyst__clear(list);
  list->alloc_idx = idx;
  return list;
}

static void
wipe_lyst(const char *file, int line, Lyst list, int destroy)
{
  LystElt cur;
  LystElt next;
  int alloc_idx;

  if (list == NULL)
  {
	  return;
  }

  alloc_idx = list->alloc_idx;
  for (cur = list->first; cur != NULL; cur = next)
  {
     next = cur->next;
     if (list->delete_cb != NULL) list->delete_cb(cur,list->delete_arg);
     lyst__elt_clear(cur); /* just in case user mistakenly accesses later... */
     lyst__free(file, line, alloc_idx, (char *) cur);
  }

  if (destroy)
  {
     lyst__clear(list); /* just in case user mistakenly accesses later... */
     lyst__free(file, line, alloc_idx, (char *) list);
  }
  else
  {
     list->first = NULL;
     list->last = NULL;
     list->length = 0;
  }
}

void
Lyst_clear(const char *file, int line, Lyst list)
{
  wipe_lyst(file, line, list, 0);
}

void
Lyst_destroy(const char *file, int line, Lyst list)
{
  wipe_lyst(file, line, list, 1);
}

/*
 * public functions - get and set list information
 */

void
lyst_compare_set(Lyst list, LystCompareFn fn)
{
  if (list != NULL) list->compare = fn;
}

LystCompareFn
lyst_compare_get(Lyst list)
{
  return (list == NULL) ? NULL : list->compare;
}

void
lyst_direction_set(Lyst list, LystSortDirection dir)
{
  if (list != NULL) list->dir = dir;
}

void
lyst_delete_set(Lyst list, LystCallback fn, void *arg)
{
  if (list != NULL)
  {
	list->delete_cb = fn;
	list->delete_arg = arg;
  }
}

void
lyst_delete_get(Lyst list, LystCallback *fn, void **arg)
{
  if (list != NULL)
  {
	CHKVOID(fn);
	CHKVOID(arg);
	*fn = list->delete_cb;
	*arg = list->delete_arg;
  }
}

void
lyst_insert_set(Lyst list, LystCallback fn, void *arg)
{
  if (list != NULL)
  {
	list->insert_cb = fn;
	list->insert_arg = arg;
  }
}

void
lyst_insert_get(Lyst list, LystCallback *fn, void **arg)
{
  if (list != NULL)
  {
	CHKVOID(fn);
	CHKVOID(arg);
	*fn = list->insert_cb;
	*arg = list->insert_arg;
  }
}

size_t
lyst_length(Lyst list)
{
  return (list == NULL) ? 0 : list->length;
}

/*
 * public functions -- add and delete elements
 */

LystElt
Lyst_insert(const char *file, int line, Lyst list, void *data)
{
  LystElt cur;

  CHKNULL(list);

  /* if not a sorted list, then just append to the end of the lyst */
  if (list->compare == NULL)
  {
      return (list->dir == LIST_SORT_ASCENDING) ?
          Lyst_insert_last(file, line, list, data) :
          Lyst_insert_first(file, line, list, data);
  }

  /* find position to insert new data into lyst */
  /* start from end of lyst to keep sort stable */
  /* because lyst__inorder returns true when both elements are equal */
  for (cur = list->last; cur != NULL; cur = cur->prev)
     if (lyst__inorder(list,cur->data,data)) break;

  /* insert into lyst */
  if (cur == NULL)
     return Lyst_insert_first(file, line, list, data);
  else
     return Lyst_insert_after(file, line, cur, data);
}

LystElt
Lyst_insert_first(const char *file, int line, Lyst list, void *data)
{
  LystElt new_elt;

  CHKNULL(list);

  /* create new element */
  if ((new_elt = lyst__elt_create(file, line, list, data)) == NULL) return NULL;

  /* insert new element at the beginning of the lyst */
  new_elt->next = list->first;
  if (list->first != NULL)
     list->first->prev = new_elt;
  else
     list->last = new_elt;
  list->first = new_elt;

  list->length += 1;

  if (list->insert_cb != NULL) list->insert_cb(new_elt,list->insert_arg);
  return new_elt;
}

LystElt
Lyst_insert_last(const char *file, int line, Lyst list, void *data)
{
  LystElt new_elt;

  CHKNULL(list);

  /* create new element */
  if ((new_elt = lyst__elt_create(file, line, list, data)) == NULL) return NULL;

  /* insert new element at the end of the lyst */
  new_elt->prev = list->last;
  if (list->last != NULL)
     list->last->next = new_elt;
  else
     list->first = new_elt;
  list->last = new_elt;

  list->length += 1;

  if (list->insert_cb != NULL) list->insert_cb(new_elt,list->insert_arg);
  return new_elt;
}

LystElt
Lyst_insert_before(const char *file, int line, LystElt elt, void *data)
{
  Lyst list;
  LystElt new_elt;

  CHKNULL(elt);
  list = elt->lyst;
  CHKNULL(list);

  /* create new element */
  if ((new_elt = lyst__elt_create(file, line, list, data)) == NULL) return NULL;

  /* insert new element before the specified element */
  new_elt->prev = elt->prev;
  new_elt->next = elt;
  if (elt->prev != NULL)
     elt->prev->next = new_elt;
  else
     list->first = new_elt;
  elt->prev = new_elt;
  list->length += 1;
  if (list->insert_cb != NULL) list->insert_cb(new_elt,list->insert_arg);
  return new_elt;
}

LystElt
Lyst_insert_after(const char *file, int line, LystElt elt, void *data)
{
  Lyst list;
  LystElt new_elt;

  CHKNULL(elt);
  list = elt->lyst;
  CHKNULL(list);

  /* create new element */
  if ((new_elt = lyst__elt_create(file, line, list, data)) == NULL) return NULL;

  /* insert new element after the specified element */
  new_elt->prev = elt;
  new_elt->next = elt->next;
  if (elt->next != NULL)
     elt->next->prev = new_elt;
  else
     list->last = new_elt;
  elt->next = new_elt;

  list->length += 1;

  if (list->insert_cb != NULL) list->insert_cb(new_elt,list->insert_arg);
  return new_elt;
}

void
Lyst_delete(const char *file, int line, LystElt elt)
{
  Lyst list;

  if (elt == NULL)
  {
	  return;
  }

  list = elt->lyst;
  CHKVOID(list);
  CHKVOID(list->length > 0);
  if (list->delete_cb != NULL) list->delete_cb(elt,list->delete_arg);

  /* update previous pointers */
  if (elt->prev != NULL)
     elt->prev->next = elt->next;
  else
     list->first = elt->next;

  /* update following pointers */
  if (elt->next != NULL)
     elt->next->prev = elt->prev;
  else
     list->last = elt->prev;

  /* free memory associated with this element */
  lyst__elt_clear(elt); /* just in case user accesses later... */
  lyst__free(file, line, list->alloc_idx, (char *) elt);

  list->length -= 1;
}

/*
 * public functions -- traverse lysts
 */

LystElt
lyst_first(Lyst list)
{
	CHKNULL(list);
	return list->first;
}

LystElt
lyst_last(Lyst list)
{
	CHKNULL(list);
	return list->last;
}

LystElt
lyst_next(LystElt elt)
{
	CHKNULL(elt);
	return elt->next;
}

LystElt
lyst_prev(LystElt elt)
{
	CHKNULL(elt);
	return elt->prev;
}

LystElt
lyst_search(LystElt elt, void *data)
{
  Lyst list;
  LystElt cur;

  CHKNULL(elt);
  list = elt->lyst;
  CHKNULL(list);

  /* linearly search lyst */
  /* should check sorted field and bail early if possible */
  for (cur = elt; cur != NULL; cur = cur->next)
  {
     if (list->compare == NULL)
     {
        /* use "==" since no comparison function provided */
        if (cur->data == data) break;
     }
     else
     {
        /* use provided comparison function */
        if (list->compare(data,cur->data) == 0) break;
     }
  }

  return cur;
}

/*
 * public functions - get and set element information
 */

Lyst
lyst_lyst(LystElt elt)
{
	CHKNULL(elt);
	return elt->lyst;
}

void *
lyst_data(LystElt elt)
{
	CHKNULL(elt);
	return elt->data;
}

void *
lyst_data_set(LystElt elt, void *new)
{
  void *old;
  Lyst list;

  CHKNULL(elt);
  list = elt->lyst;
  CHKNULL(list);
  old = elt->data;
  elt->data = new;
  return old;
}

/*
 * public functions -- miscellaneous
 */

void
lyst_sort(Lyst list)
{
  LystElt cur;
  LystElt next;
  LystElt elt;

  CHKVOID(list);
  if (list->compare == NULL)
  {
	  return;
  }

  for (cur = list->first; cur != NULL; cur = next)
  {
     next = cur->next;

     /* find place for cur element in sorted part of lyst */
     /* start from end of lyst to keep sort stable */
     /* because lyst__inorder returns true when both elements are equal */
     for (elt = cur->prev; elt != NULL; elt = elt->prev)
        if (lyst__inorder(list,elt->data,cur->data)) break;

     /* check to see if cur element was already in correct position */
     if (elt != cur->prev)
     {
        /* remove cur element from lyst */
        /* there is always at least one element before cur */
        cur->prev->next = cur->next;
        if (cur->next == NULL)
           list->last = cur->prev;
        else
           cur->next->prev = cur->prev;

        /* insert cur element after elt element */
        /* there is always at least one element after elt */
        cur->prev = elt;
        if (elt == NULL)
        {
           cur->next = list->first;
           list->first = cur;
        }
        else
        {
           cur->next = elt->next;
           elt->next = cur;
        }
        cur->next->prev = cur;
     }
  }
}

int
lyst_sorted(Lyst list)
{
  int sorted;
  LystElt cur;

  CHKZERO(list);
  if (list->compare == NULL)
  {
	  return 0;
  }

  sorted = 1;
  if ((cur = list->first) != NULL)
  {
     while (cur->next != NULL)
     {
        if (!lyst__inorder(list,cur->data,cur->next->data)) break;
        cur = cur->next;
     }
     if (cur->next != NULL) sorted = 0;
  }

  return sorted;
}

void
lyst_apply(Lyst list, LystCallback fn, void *user_arg)
{
  LystElt cur;
  LystElt next;

  CHKVOID(list);
  CHKVOID(fn);
  for (cur = list->first; cur != NULL; cur = next)
  {
     next = cur->next;
     (*fn)(cur,user_arg);
  }
}

/*
 * private functions -- zero out/initialize fields of a lyst or element
 */

static void
lyst__clear(Lyst list)
{
  list->first = NULL;
  list->last = NULL;
  list->length = 0;
  list->compare = NULL;
  list->dir = LIST_SORT_ASCENDING;
  list->delete_cb = NULL;
  list->delete_arg = NULL;
  list->insert_cb = NULL;
  list->insert_arg = NULL;
  list->alloc_idx = 0;
}

static int
lyst__inorder(Lyst list, void *data1, void *data2)
{
  return (list->dir == LIST_SORT_ASCENDING && list->compare(data1,data2) <= 0)
  || (list->dir == LIST_SORT_DESCENDING && list->compare(data1,data2) >= 0);
}

static LystElt
lyst__elt_create(const char *file, int line, Lyst list, void *data)
{
  LystElt elt;

  if ((elt = (LystElt) lyst__alloc(file, line, list->alloc_idx, sizeof(*elt)))
	== NULL)
  {
	putErrmsg("Can't create list element.", NULL);
	return NULL;
  }

  lyst__elt_clear(elt);
  elt->lyst = list;
  elt->data = data;

  return elt;
}

static void
lyst__elt_clear(LystElt elt)
{
  elt->lyst = NULL;
  elt->prev = NULL;
  elt->next = NULL;
  elt->data = NULL;
}

static char *lyst__alloc(const char *fileName, int lineNbr, int idx,
		unsigned size)
{
  MemAllocator	take = memmgr_take(idx);
  char		*ptr;

  CHKNULL(take);
  ptr = take(fileName, lineNbr, size);
  if (ptr == NULL)
  {
	  putErrmsg("Lyst memory allocation failed.", utoa(size));
  }

  return ptr;
}

static void lyst__free(const char *fileName, int lineNbr, int idx, char *ptr)
{
  MemDeallocator	release = memmgr_release(idx);

  if (release != NULL) release(fileName, lineNbr, (void *) ptr);
}
